sale_rental/models/sale_order.py

```diff
@@ -0,0 +1,300 @@
+                    raise ValidationError(_(
+                        "On the sale order line with product '%s' "
+                        "the Product Quantity (%s) should be the "
+                        "number of days (%s) "
+                        "multiplied by the Rental Quantity (%s).") % (
+                        line.product_id.name, line.product_uom_qty,
+                        line.number_of_days, line.rental_qty))
+                # the module sale_start_end_dates checks that, when we have
+                # must_have_dates, we have start + end dates
+            elif line.sell_rental_id:
+                if line.product_uom_qty != line.sell_rental_id.rental_qty:
+                    raise ValidationError(_(
+                        "On the sale order line with product %s "
+                        "you are trying to sell a rented product with a "
+                        "quantity (%s) that is different from the rented "
+                        "quantity (%s). This is not supported.") % (
+                        line.product_id.name,
+                        line.product_uom_qty,
+                        line.sell_rental_id.rental_qty))
+
+    @api.multi
+    def _prepare_rental(self):
+        self.ensure_one()
+        return {'start_order_line_id': self.id}
+
+    @api.multi
+    def _get_rental_date_planned(self):
+        self.ensure_one()
+        start_date = datetime.strptime(
+            self.start_date, DEFAULT_SERVER_DATE_FORMAT)
+        return start_date.strftime(DEFAULT_SERVER_DATETIME_FORMAT)
+
+    @api.multi
+    def _action_launch_procurement_rule(self):
+        res = super(SaleOrderLine, self)._action_launch_procurement_rule()
+        errors = []
+        for line in self:
+            if not line.product_id.rented_product_id or\
+                    line.rental_type == 'rental_extension':
+                continue
+
+            procurement_group = line.order_id.procurement_group_id
+            if not procurement_group:
+                procurement_group = self.env['procurement.group'].create({
+                    'name': line.order_id.name,
+                    'move_type': line.order_id.picking_policy,
+                    'sale_id': line.order_id.id,
+                    'partner_id': line.order_id.partner_shipping_id.id,
+                })
+                line.order_id.procurement_group_id = procurement_group
+
+            if line.rental_type == 'new_rental':
+                vals = {
+                    'company_id': line.order_id.company_id,
+                    'group_id': procurement_group,
+                    'sale_line_id': line.id,
+                    'date_planned': line._get_rental_date_planned(),
+                    'route_ids': line.order_id.warehouse_id.rental_route_id,
+                    'warehouse_id': line.order_id.warehouse_id or False,
+                    'partner_dest_id': line.order_id.partner_shipping_id
+                }
```

Comment:
Could you please move this part into a new function of "sale.order.line", so that it can be extended easily.